fem/lininteg_domain.cpp

```diff
@@ -0,0 +1,339 @@
+      const DeviceMatrix Bt(sBt, d,q);
+      kernels::internal::LoadB<D,Q>(d,q,B,sBt);
+
+      MFEM_SHARED double sQQQ[Q*Q*Q];
+      const DeviceCube QQQ(sQQQ, q,q,q);
+
+      for (int c = 0; c < vdim; ++c)
+      {
+         const double cst_val = C(c,0,0,0,0);
+         MFEM_FOREACH_THREAD(x,x,q)
+         {
+            MFEM_FOREACH_THREAD(y,y,q)
+            {
+               for (int z = 0; z < q; ++z)
+               {
+                  const double J11 = J(x,y,z,0,0,e);
+                  const double J21 = J(x,y,z,1,0,e);
+                  const double J31 = J(x,y,z,2,0,e);
+                  const double J12 = J(x,y,z,0,1,e);
+                  const double J22 = J(x,y,z,1,1,e);
+                  const double J32 = J(x,y,z,2,1,e);
+                  const double J13 = J(x,y,z,0,2,e);
+                  const double J23 = J(x,y,z,1,2,e);
+                  const double J33 = J(x,y,z,2,2,e);
+                  const double detJ = J11 * (J22 * J33 - J32 * J23) -
+                  /* */               J21 * (J12 * J33 - J32 * J13) +
+                  /* */               J31 * (J12 * J23 - J22 * J13);
+                  const double coeff_val = cst_coeff ? cst_val : C(c,x,y,z,e);
+                  QQQ(z,y,x) = W(x,y,z) * coeff_val * detJ;
+               }
+            }
+         }
+         MFEM_SYNC_THREAD;
+         MFEM_FOREACH_THREAD(qx,x,q)
+         {
+            MFEM_FOREACH_THREAD(qy,y,q)
+            {
+               for (int dz = 0; dz < d; ++dz) { u[dz] = 0.0; }
+               for (int qz = 0; qz < q; ++qz)
+               {
+                  const double ZYX = QQQ(qz,qy,qx);
+                  for (int dz = 0; dz < d; ++dz) { u[dz] += ZYX * Bt(dz,qz); }
+               }
+               for (int dz = 0; dz < d; ++dz) { QQQ(dz,qy,qx) = u[dz]; }
+            }
+         }
+         MFEM_SYNC_THREAD;
+         MFEM_FOREACH_THREAD(dz,y,d)
+         {
+            MFEM_FOREACH_THREAD(qx,x,q)
+            {
+               for (int dy = 0; dy < d; ++dy) { u[dy] = 0.0; }
+               for (int qy = 0; qy < q; ++qy)
+               {
+                  const double zYX = QQQ(dz,qy,qx);
+                  for (int dy = 0; dy < d; ++dy) { u[dy] += zYX * Bt(dy,qy); }
+               }
+               for (int dy = 0; dy < d; ++dy) { QQQ(dz,dy,qx) = u[dy]; }
+            }
+         }
+         MFEM_SYNC_THREAD;
+         MFEM_FOREACH_THREAD(dz,y,d)
+         {
+            MFEM_FOREACH_THREAD(dy,x,d)
+            {
+               for (int dx = 0; dx < d; ++dx) { u[dx] = 0.0; }
+               for (int qx = 0; qx < q; ++qx)
+               {
+                  const double zyX = QQQ(dz,dy,qx);
+                  for (int dx = 0; dx < d; ++dx) { u[dx] += zyX * Bt(dx,qx); }
+               }
+               for (int dx = 0; dx < d; ++dx) { Y(dx,dy,dz,c,e) += u[dx]; }
+            }
+         }
+         MFEM_SYNC_THREAD;
+      }
+   });
+}
+
+static void LaunchDeviceKernel(const FiniteElementSpace &fes,
```

Comment:
Maybe this should have a less generic name (or be namespaced somehow), it's launching a very specific device kernel.

Reply:
Renamed to `DLF[Eval|Grad]Assemble`